lib/posix/pthread.c

```diff
@@ -36,23 +56,68 @@ static const struct pthread_attr init_pthread_attrs = {
 	.initialized = true,
 };
 
-static struct posix_thread posix_thread_pool[CONFIG_MAX_PTHREAD_COUNT];
-static struct k_spinlock pthread_pool_lock;
+/*
+ * We reserve the MSB to mark a pthread_t as initialized (from the
+ * perspective of the application). With a linear space, this means that
+ * the theoretical pthread_t range is [0,2147483647].
+ */
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_COUNT < PTHREAD_OBJ_MASK_INIT,
+	     "CONFIG_MAX_PTHREAD_COUNT is too high");
 
-pthread_t pthread_self(void)
+static inline size_t posix_thread_to_offset(struct posix_thread *t)
+{
+	return t - posix_thread_pool;
+}
+
+static inline size_t get_posix_thread_idx(pthread_t pth)
 {
-	return (struct posix_thread *)
-		CONTAINER_OF(k_current_get(), struct posix_thread, thread)
-		- posix_thread_pool;
+	return mark_pthread_obj_uninitialized(pth);
 }
 
-struct posix_thread *to_posix_thread(pthread_t pthread)
+struct posix_thread *to_posix_thread(pthread_t pth)
 {
-	if (pthread >= CONFIG_MAX_PTHREAD_COUNT) {
+	k_spinlock_key_t key;
+	struct posix_thread *t;
+	bool actually_initialized;
+	size_t bit = get_posix_thread_idx(pth);
+
+	/* if the provided thread does not claim to be initialized, its invalid */
+	if (!is_pthread_obj_initialized(pth)) {
+		return NULL;
+	}
+
+	if (bit >= CONFIG_MAX_PTHREAD_COUNT) {
+		return NULL;
+	}
+
+	t = &posix_thread_pool[bit];
+
+	key = k_spin_lock(&q_lock);
+	/*
+	 * Denote a pthread as "initialized" (i.e. allocated) if it is not in ready_q.
+	 * This differs from other posix object allocation strategies because they use
+	 * a bitarray to indicate whether an object has been allocated.
+	 */
+	actually_initialized = t->qid != POSIX_THREAD_READY_Q;
+	k_spin_unlock(&q_lock, key);
+
+	if (actually_initialized == 0) {
+		/* The thread claims to be initialized but is actually not */
 		return NULL;
 	}
 
-	return &posix_thread_pool[pthread];
+	return &posix_thread_pool[bit];
+}
+
+pthread_t pthread_self(void)
+{
+	size_t bit;
+	struct posix_thread *t;
+
+	t = (struct posix_thread *)CONTAINER_OF(k_current_get(), struct posix_thread, thread);
```

Comment:
Would it be good to have an assert that the current thread was actually a posix thread? There's no valid error return from this function, so it would be a fatal error.

Reply:
Oddly enough, we need to be able to call `pthread_self()` even with non-posix threads for comparison purposes.